Sources/XiEditor/MarkerBar.swift

```diff
@@ -0,0 +1,89 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Cocoa
+
+class Marker {
+    let line: Int
+    let color: NSColor
+
+    init(_ line: Int, color: NSColor) {
+        self.line = line
+        self.color = color
+    }
+
+    @available(*, unavailable)
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+}
+
+class MarkerBar: NSScroller {
+    var markers: [Marker] = []
+    var backgroundColor: NSColor = NSColor.red
+    var parent: EditViewController!
+    var markerLayer = CAShapeLayer()
+    var overlayScrollerLayer: CALayer?
+
+    override func viewWillDraw() {
+        super.viewWillDraw()
+
+        // last child layer will be used for drawing the overlay scroller
+        overlayScrollerLayer = self.layer?.sublayers?.last
+        overlayScrollerLayer?.addObserver(self, forKeyPath: "opacity", options: .new, context: nil)
+        self.layer?.addSublayer(markerLayer)
+    }
+
+    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
+        if keyPath == "opacity" {
+            markerLayer.opacity = change![NSKeyValueChangeKey.newKey] as! Float
+        }
+    }
+
+    var markerHeight: CGFloat = 1.0
+
+    override class var isCompatibleWithOverlayScrollers: Bool {
+        return true
+    }
+
+    override func drawKnob() {
+        drawMarkers()
+        super.drawKnob()
+    }
+
+    func setMarker(_ items: [Marker]) {
+        markers = items
+        drawMarkers()
+    }
+
+    func drawMarkers() {
+        let path = CGMutablePath()
+        let totalLines = CGFloat(parent.lines.height)
+        let markerBarHeight = self.layer?.bounds.height
+        let maxScrollerWidth = self.layer?.bounds.width
```

Comment:
I see a few different approaches with these optional values.

The first question to ask with an `Option` type (here or in rust) is, "can this value ever be (`None`/`nil`) in normal use?" If the answer is 'definitely not', then the best thing to do is to unwrap immediately, and then the rest of the function doesn't have to deal with them.

If the answer is "maybe", or "I don't know", or "yes", then the second question is, "Is this function still meaningful if this value is (`None`/`nil`)?" if yes, then you want to provide a default value in the `nil` case, and if not you want to exit early, maybe logging that something unexpected happened.

More concretely, with this function it is _unlikely_ that the layer doesn't exist when this gets called, but I'm not sure it's impossible. It's worth at least being defensive. As this function _is not_  meaningful in that case, we should exit early. In swift, the pattern for this is a guard statement:

```swift
    func drawMarkers() {
        guard let markerBarHeight = layer?.bounds.height,
            let maxScrollerWidth = layer?.bounds.width,
            let width = overlayScrollerLayer?.bounds.width else { return; }
    // no more unwraps required
```